### PR TITLE
Add constbuffer_array module

### PR DESCRIPTION
**Pull request by dcristoloveanu**
Add constbuffer_array module

### CODE REVIEW

File: devdoc/constbuffer_array_requirements.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+/*remove front*/
+MOCKABLE_FUNCTION(, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_remove_front, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_handle, CONSTBUFFER_HANDLE *const_buffer_handle);
+
+/* getters */
+MOCKABLE_FUNCTION(, int, constbuffer_array_get_buffer_count, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_handle, uint32_t*, buffer_count);
+MOCKABLE_FUNCTION(, CONSTBUFFER_HANDLE, constbuffer_array_get_buffer, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_handle, uint32_t, buffer_index);
+```
+
+### constbuffer_array_create_empty
```
Comment (anporumb) [inline devdoc/constbuffer_array_requirements.md]: >constbuffer_array_create_empty [](start = 4, length = 30)

constbuffer_array_create

File: devdoc/constbuffer_array_requirements.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+**SRS_CONSTBUFFER_ARRAY_01_009: [** `constbuffer_array_create` shall allocate memory for a new `CONSTBUFFER_ARRAY_HANDLE` that can hold `buffer_count` buffers. **]**
+
+**SRS_CONSTBUFFER_ARRAY_01_010: [** `constbuffer_array_create` shall clone the buffers in `buffers` and store them. **]**
+
+**SRS_CONSTBUFFER_ARRAY_01_011: [** On success `constbuffer_array_create` shall return a non-NULL handle. **]**
+
+**SRS_CONSTBUFFER_ARRAY_01_012: [** If `buffers` is NULL, `constbuffer_array_create` shall fail and return NULL. **]**
+
+**SRS_CONSTBUFFER_ARRAY_01_013: [** If `buffer_count` is 0, `constbuffer_array_create` shall fail and return NULL. **]**
```
Comment (anporumb) [inline devdoc/constbuffer_array_requirements.md]: >buffer_count [](start = 40, length = 12)

consider allowing NULL for buffers when the count is 0.

File: devdoc/constbuffer_array_requirements.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+```c
+MOCKABLE_FUNCTION(, CONSTBUFFER_ARRAY_HANDLE, constbuffer_array_create_empty);
+```
+
+`constbuffer_array_create_empty` creates a new, empty CONSTBUFFER_ARRAY_HANDLE.
+
+**SRS_CONSTBUFFER_ARRAY_02_004: [** `constbuffer_array_create_empty` shall allocate memory for a new `CONSTBUFFER_ARRAY_HANDLE`. **]**
+
+**SRS_CONSTBUFFER_ARRAY_02_005: [** `constbuffer_array_create_empty` shall maintain an internal array of `CONSTBUFFER_HANDLE`s **]**
```
Comment (anporumb) [inline devdoc/constbuffer_array_requirements.md]: >maintain an internal array [](start = 75, length = 26)

If it is empty (line 58 says so), then maybe it doesn't need to maintain it...

File: devdoc/constbuffer_array_requirements.md (hunk 1 above)
Comment (dcristoloveanu) [reply devdoc/constbuffer_array_requirements.md]: Fixed.

---
In reply to: [234323059](https://github.com/Azure/azure-c-shared-utility/pull/226#discussion_r234323059) [](ancestors = 234323059)

File: devdoc/constbuffer_array_requirements.md (hunk 3 above)
Comment (dcristoloveanu) [reply devdoc/constbuffer_array_requirements.md]: Removed spec.

---
In reply to: [234323937](https://github.com/Azure/azure-c-shared-utility/pull/226#discussion_r234323937) [](ancestors = 234323937)

File: devdoc/constbuffer_array_requirements.md (hunk 2 above)
Comment (dcristoloveanu) [reply devdoc/constbuffer_array_requirements.md]: Ok. Will do

---
In reply to: [234323603](https://github.com/Azure/azure-c-shared-utility/pull/226#discussion_r234323603) [](ancestors = 234323603)

**Pull request merged**